### ISSUE TITLE
SequelizeUniqueConstraintError on child creation

### ISSUE THREAD

**Issue by Varkal** [overlookmotel/sequelize-hierarchy#65]
This simple test throw a SequelizeUniqueConstraintError

```
beforeEach((done) => {
            models.sequelize.sync({force: true}).then(() => {
                this.Account = models.Account;
                done();
            });
});

it('shoud handle correctly hierarchy of account', () => {
            var parentAccount = {
                title  : "Parent Account",
                balance: 0
            };

            var childAccount = {
                title  : "Child Account",
                balance: 0
            };

            return this.Account.create(parentAccount)
                .then((parentAccountCreated) => {
                    childAccount.parentId = parentAccountCreated.id;
                    return this.Account.create(childAccount)
                })
                .then(
                    (childAccountCreated) => {
                        console.log(childAccountCreated.parentId);
                    },
                    (error) => {
                        console.log(error);
                    }
                )
        });
```
But, after exec it, in MySQL i dound those datas : 

| id | title          | balance | deleted | deletion_date | creation_date       | modification_date   | hierarchyLevel | parentId |
+----+----------------+---------+---------+---------------+---------------------+---------------------+-------------+
|  1 | Parent Account |       0 |    NULL | NULL          | 2016-06-25 13:51:35 | 2016-06-25 13:51:35 |              1 |     NULL |
|  2 | Child Account  |       0 |    NULL | NULL          | 2016-06-25 13:51:35 | 2016-06-25 13:51:35 |              2 |        1 |

Like everything is ok.

So i don't understand where is my mistake...

Do you have any idea ? 

Some additional infos : 

The detailed error : 
```
{ [SequelizeUniqueConstraintError: Validation error]
  name: 'SequelizeUniqueConstraintError',
  message: 'Validation error',
  errors:
   [ { message: 'PRIMARY must be unique',
       type: 'unique violation',
       path: 'PRIMARY',
       value: '2-1' } ],
  fields: { PRIMARY: '2-1' } }
```
The correspondig model : 
```
var Account = db.define("Account", {
            title        : {
                type: Sequelize.STRING
            },
            balance      : {
                type: Sequelize.FLOAT
            },
            deleted      : {
                type: Sequelize.BOOLEAN,
            },
            deletion_date: {
                type: Sequelize.DATE
            }
        },
        {
            timestamps  : true,
            createdAt   : "creation_date",
            updatedAt   : "modification_date",
            hierarchy   : true
        });

    return Account;
```

**Comment by overlookmotel** [comment]
Thanks for raising this.

When you set a hierarchy on a model, it also defines another model to track the descendents of all nodes. In your case this would be called `AccountHierarchy`.

It looks like the error is coming from trying to insert a duplicated row into `AccountHierarchy` table.

Two possibilities for why this is happening:

1. A bug in this module
2. `AccountHierarchy` table is not being cleared before the test run so row inserted in a previous test or previous run of this test are still present.

I suspect it's (2) as there are already tests for this which are passing fine.

To narrow this down, I'd suggest the following:

* Before running that test, manually delete or empty the `AccountHierarchy` table from the database, and see if you still get the same error.
* Ensure that `models.sequelize.sync({force: true})` is syncing *all* models, including `AccountHierarchy`. From your code, I can't tell whether `models.sequelize` refers to the the `db` object that `Account` model is being defined on.

Please can you try this and let me know? If there's a bug, of course I want to fix it ASAP, but I *suspect* the fault is with your code somewhere.

**Comment by overlookmotel** [comment]
What version of node and sequelize are you using?

**Comment by Varkal** [comment]
Thanks for the quick answer ! 

So : 

node v4.4.6
sequelize": "^3.23.4",
sequelize-hierarchy": "^0.7.0"

I've tried to comment all my tests, except of the one which bugs.

Schema in DB : 
```
+---------------------------+
| Tables_in_account_manager |
+---------------------------+
| Accounts                  |
| Accountsancestors   |
+---------------------------+
```

Manually delete all datas;
```
mysql> SELECT * FROM Accounts;
Empty set (0.00 sec)
```

```
mysql> SELECT * FROM Accountsancestors;
Empty set (0.00 sec)
```
And the first time i run the tests after, it's ok.

So i think it's a problem with my code...

I've followed the example i've found here to organize my app : https://github.com/sequelize/express-example/

So my models file is : 

```
//See http://docs.sequelizejs.com/en/1.7.0/articles/express/

var fs        = require("fs");
var path      = require("path");
var Sequelize = require("sequelize");
var config    = require("../config/config.json");

//Sequelize plugins
require('sequelize-hierarchy')(Sequelize);


var sequelize = new Sequelize(config.database, config.username, config.password, config);
var db        = {};

fs
    .readdirSync(__dirname)
    .filter((file) => {
        return (file.indexOf(".") !== 0) && (file !== "index.js");
    })
    .forEach((file) => {
        var model = sequelize.import(path.join(__dirname, file));
        db[model.name] = model;
    });

Object.keys(db).forEach((modelName) => {
    if ("associate" in db[modelName]) {
        db[modelName].associate(db);
    }
});

db.sequelize = sequelize;
db.Sequelize = Sequelize;

module.exports = db;
```
And so `models.sequelize.sync({force:true})` shoud sync all my models. (If i understand how Sequelize works)
 
 Thanks for the help !

**Comment by overlookmotel** [comment]
I've put all the code you've posted into a little repo and run the tests but I'm unable to reproduce the fault.

You are not by any chance doing `require('sequelize-hierarchy')(Sequelize)` in another file as well?

`require('sequelize-hierarchy')(Sequelize)` adds hooks on sequelize's methods like `model.create()`. So if you initialize `sequelize-hierarchy` twice, it adds the hooks twice. So then it will try to add the same rows to `Accountsancestors` table twice, which causes a primary key unique error.

I added a 2nd `require('sequelize-hierarchy')(Sequelize)` to my test repo and got exactly the same error as you.

Please let me know if that solves it...

**Comment by Varkal** [comment]
No, it's require only once...

I try to replace mysql with sqlite, but i have the same problem

I add nother beforeEach block, to be sure : 

```
var chai   = require('chai').should();
var models = require("../../src/models");

describe("The Account model", function (){
    beforeEach(function(done){
        models.sequelize.sync({force: true}).then(() => {
            this.Account = models.Account;
            done();
        });
    });


    describe("The create method", function(){
        beforeEach(function(done){
            models.sequelize.query("SELECT * FROM Accountsancestors WHERE 1", {type: models.Sequelize.QueryTypes.SELECT})
                .then((result, metadata) => {
                    console.log(result);
                    done();
                });
        });

        it('shoud return the account created', function(done){
            var testAccount = {
                title  : "Test Account",
                balance: 0
            };

            return this.Account.create(testAccount).then((account) => {
                account.title.should.equal(testAccount.title);
                done();
            })
        });

        it('shoud add the account created to the database', function(done){
            var testAccount = {
                title  : "Test Account",
                balance: 0
            };

            return this.Account.create(testAccount).then((accountCreated) => {
                return this.Account.findById(accountCreated.id).then((accountFound) => {
                    accountFound.title.should.equal(testAccount.title);
                    done();
                })
            })
        });

        it('shoud handle correctly hierarchy of account', function(done){
            var parentAccount = {
                title  : "Parent Account",
                balance: 0
            };

            var childAccount = {
                title  : "Child Account",
                balance: 0
            };


            return this.Account.create(parentAccount)
                .then((parentAccountCreated) => {
                    childAccount.parentId = parentAccountCreated.id;
                    return this.Account.create(childAccount)
                })
                .then(
                    (childAccountCreated) => {
                        childAccountCreated.parentId.should.equal(1);
                        done();
                    },
                    (error) => {
                        console.log(error);
                        done();
                    }
                )
        });
    });
});
```

And the result : Works on first start, but failed after

  ```
The Account model
    The create method
[]
      ✓ shoud return the account created
[]
      ✓ shoud add the account created to the database
[]
      ✓ shoud handle correctly hierarchy of account


  3 passing (251ms)



  The Account model
    The create method
[]
      ✓ shoud return the account created
[]
      ✓ shoud add the account created to the database
[]
{ [SequelizeUniqueConstraintError: Validation error]
  name: 'SequelizeUniqueConstraintError',
  message: 'Validation error',
  errors:
   [ { message: 'PRIMARY must be unique',
       type: 'unique violation',
       path: 'PRIMARY',
       value: '2-1' } ],
  fields: { PRIMARY: '2-1' } }
      ✓ shoud handle correctly hierarchy of account


  3 passing (156ms)
```

Once again, thanks for your time and your help.

**Comment by overlookmotel** [comment]
OK, this is really mysterious. I can't replicate this.

So I'm sure I'm not missing anything, could you please put the whole app (well the relevant files anyway) in a Github repo? Obviously remove the db authentication credentials from `config.json` but please keep `config.json` in the repo so I can check there isn't something in there that's part of this problem.

**Comment by overlookmotel** [comment]
By the way, if you remove the last then handler:

```js
(error) => {
    console.log(error);
    done();
}
```

Then the error will be thrown back to mocha, will cause the test to fail, and it'll print the stack trace. Which could be useful in locating the problem.

**Comment by Varkal** [comment]
Sure, i can share the code : https://github.com/Varkal/Temp

Thanks you so much it's so kind to spend so much time to help me.

**Comment by overlookmotel** [comment]
Ok. I got to the bottom of it.

There is something weird going on with mocha when using the `--watch` option is used.

Add the following lines to `src/models/index.js` just before `require('sequelize-hierarchy')(Sequelize)`:

```js
console.log('Attaching sequelize-hierarchy');
if (Sequelize.HierarchyError) console.log('sequelize-hierarchy already attached!');
```

When you run the tests the first time, you'll just see `Attaching sequelize-hierarchy`. When you alter a file which triggers `mocha` to run again, you'll see in the output `sequelize-hierarchy already attached!`.

That's what's causing the problem - sequelize-hierarchy is being attached twice, which causes it to run its hooks twice, which means it tries to insert the same rows twice. And that's what causes the error.

I'd suggest not using `mocha`'s `--watch` option. Or raise an issue with `mocha`.

Also, I noticed that your tests are implemented a bit oddly - you're returning promises and using callbacks simultaneously. If you return promises, you don't need the `done` callbacks, and in some cases doing so is causing errors to get swallowed. I've changed the test functions to all use just promises here: https://github.com/Varkal/Temp/pull/1/commits/bd3d61345e29693bd103dd879fd07661b73a484d

I conclude that there's no bug in sequelize-hierarchy, so closing this issue.
